File: AgoraChat/Code/Controllers/Contacts/ACDRequestListViewController.m
//
//  AgoraRequestListViewController.m
//  ChatDemo-UI3.0
//
//  Created by liang on 2021/10/21.
//  Copyright © 2021 easemob. All rights reserved.
//

#import "ACDRequestListViewController.h"
#import "MISScrollPage.h"
#import "ACDRequestCell.h"
#import "AgoraApplyManager.h"
#import "AgoraApplyModel.h"

@interface ACDRequestListViewController ()

@end

@implementation ACDRequestListViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.view.backgroundColor = UIColor.whiteColor;
}


- (void)updateUI {
    [self.dataArray removeAllObjects];
    
    NSArray *contactApplys = [[AgoraApplyManager defaultManager] contactApplys];
    NSArray *groupApplys = [[AgoraApplyManager defaultManager] groupApplys];
    [self.dataArray addObjectsFromArray:contactApplys];
    [self.dataArray addObjectsFromArray:groupApplys];
    
    self.searchSource = [NSMutableArray arrayWithArray:self.dataArray];
    [self.table reloadData];
}

#pragma mark action
- (void)declineAction:(AgoraApplyModel *)model {
    WEAK_SELF
    [MBProgressHUD showHUDAddedTo:[UIApplication sharedApplication].keyWindow animated:YES];
    switch (model.style) {
        case AgoraApplyStyle_contact:
        {
            [[AgoraChatClient sharedClient].contactManager declineFriendRequestFromUser:model.applyHyphenateId completion:^(NSString *aUsername, AgoraChatError *aError) {
                [weakSelf declineApplyFinished:aError model:model];
            }];

            break;
        }
        case AgoraApplyStyle_joinGroup:
        {
            [[AgoraChatClient sharedClient].groupManager declineJoinGroupRequest:model.groupId sender:model.applyHyphenateId reason:nil completion:^(AgoraChatGroup *aGroup, AgoraChatError *aError) {
                [weakSelf declineApplyFinished:aError model:model];
            }];
            break;
        }
        default:
        {
            [[AgoraChatClient sharedClient].groupManager declineGroupInvitation:model.groupId inviter:model.applyHyphenateId reason:nil completion:^(AgoraChatError *aError) {
                [weakSelf declineApplyFinished:aError model:model];
            }];
            break;
        }
    }
}


- (void)acceptAction:(AgoraApplyModel *)model {
    WEAK_SELF
    [MBProgressHUD showHUDAddedTo:[UIApplication sharedApplication].keyWindow animated:YES];
    switch (model.style) {
        case AgoraApplyStyle_contact:
        {
            [[AgoraChatClient sharedClient].contactManager approveFriendRequestFromUser:model.applyHyphenateId completion:^(NSString *aUsername, AgoraChatError *aError) {
                [weakSelf acceptApplyFinished:aError model:model];
            }];
            break;
        }
        case AgoraApplyStyle_joinGroup:
        {
            [[AgoraChatClient sharedClient].groupManager approveJoinGroupRequest:model.groupId sender:model.applyHyphenateId completion:^(AgoraChatGroup *aGroup, AgoraChatError *aError) {
                [weakSelf acceptApplyFinished:aError model:model];
            }];
            break;
        }
        default:
        {
            [[AgoraChatClient sharedClient].groupManager acceptInvitationFromGroup:model.groupId inviter:model.applyHyphenateId completion:^(AgoraChatGroup *aGroup, AgoraChatError *aError) {
                [weakSelf acceptApplyFinished:aError model:model];
            }];
            break;
        }
    }
}

- (void)declineApplyFinished:(AgoraChatError *)error
                       model:(AgoraApplyModel*)model{
    [MBProgressHUD hideAllHUDsForView:[UIApplication sharedApplication].keyWindow animated:YES];
    if (!error) {
        [[AgoraApplyManager defaultManager] removeApplyRequest:model];
    }
    else {
        [self showAlertWithMessage:@"Refused to apply for failure"];
    }
    
    [self.table reloadData];
}

- (void)acceptApplyFinished:(AgoraChatError *)error
                      model:(AgoraApplyModel*)model {
    [MBProgressHUD hideAllHUDsForView:[UIApplication sharedApplication].keyWindow animated:YES];
    if (!error) {
        [[AgoraApplyManager defaultManager] removeApplyRequest:model];
    }
    else {
        [self showAlertWithMessage:@"Failed to agree to apply"];
    }
    
    [self.table reloadData];
}


#pragma mark - Table view data source
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    if (self.isSearchState) {
        return [self.searchResults count];
    }
    return self.dataArray.count;
}


- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    static NSString *cellIdentifier = @"AgoraGroupCell";
    ACDRequestCell *cell = [tableView dequeueReusableCellWithIdentifier:cellIdentifier];
    if (!cell) {
        cell = [[ACDRequestCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:cellIdentifier];
    }
    
    AgoraApplyModel *applyModel = nil;
    if (self.isSearchState) {
        applyModel = self.searchResults[indexPath.row];
    }else {
        applyModel = self.dataArray[indexPath.row];
    }
    
    [cell updateWithObj:applyModel];
    ACD_WS
    cell.acceptBlock = ^(AgoraApplyModel * _Nonnull model) {
        applyModel.applyStatus = ACDApplyStatusAgreed;
        [weakSelf acceptAction:model];
    };
    
    cell.rejectBlock = ^(AgoraApplyModel * _Nonnull model) {
        applyModel.applyStatus = ACDApplyStatusDeclined;
        [weakSelf declineAction:model];
    };
    
    return cell;
}


#pragma mark getter and setter
- (UITableView *)table {
    if (_table == nil) {
        _table                 = [[UITableView alloc] initWithFrame:CGRectMake(0, 0, KScreenWidth, KScreenHeight) style:UITableViewStylePlain];
        _table.delegate        = self;
        _table.dataSource      = self;
        _table.separatorStyle  = UITableViewCellSeparatorStyleNone;
        _table.keyboardDismissMode = UIScrollViewKeyboardDismissModeOnDrag;
        [_table registerClass:[ACDRequestCell class] forCellReuseIdentifier:[ACDRequestCell reuseIdentifier]];
        _table.rowHeight = 102.0f;
    }
    return _table;
}


#pragma mark MISScrollPageControllerContentSubViewControllerDelegate
- (void)viewDidAppearForIndex:(NSUInteger)index{
    [self updateUI];
    [[AgoraChatDemoHelper shareHelper] hiddenApplyRedPoint];
}

- (void)viewDidDisappearForIndex:(NSUInteger)index {
    [self updateUI];
    [[AgoraChatDemoHelper shareHelper] setupUntreatedApplyCount];

}

@end

File: AgoraChat-Swift/AgoraChat-Swift/code/Controllers/Contacts/ACDRequestListViewController.swift
//
//  ACDRequestListViewController.swift
//  AgoraChat-Swift
//
//  Created by 冯钊 on 2022/10/18.
//

import UIKit

class ACDRequestListViewController: ACDContainerSearchTableViewController<AgoraApplyModel> {

    override func viewDidLoad() {
        super.viewDidLoad()
        self.view.backgroundColor = .white
        
        self.table.separatorStyle = .none
        self.table.keyboardDismissMode = .onDrag
        self.table.rowHeight = 102
        self.table.register(UINib(nibName: "ACDRequestCell", bundle: nil), forCellReuseIdentifier: "cell")
    }
    
    func updateUI() {
        self.dataArray.removeAll()
        let contactApplys = AgoraApplyManager.shared.reloadContactApplys()
        self.dataArray.append(contentsOf: contactApplys)

        let groupApplys = AgoraApplyManager.shared.reloadGroupApplys()
        self.dataArray.append(contentsOf: groupApplys)
        
        self.searchSource = self.dataArray
        self.table.reloadData()
    }
    
    private func declineApplyFinished(model: AgoraApplyModel, error: AgoraChatError?) {
        MBProgressHUD.hide(for: self.view, animated: true)
        if error == nil {
            model.status = .declined
            AgoraApplyManager.shared.addOrUpdate(apply: model)
        } else {
            self.showAlert(message: "Refused to apply for failed".localized)
        }
        self.table.reloadData()
    }
    
    private func acceptApplyFinished(model: AgoraApplyModel, error: AgoraChatError?) {
        MBProgressHUD.hide(for: self.view, animated: true)
        if error == nil {
            model.status = .agreed
            AgoraApplyManager.shared.remove(apply: model)
            let body = AgoraChatTextMessageBody(text: "You agreed the friend request".localized)
            let message = AgoraChatMessage(conversationID: model.userId, from: model.userId, to: AgoraChatClient.shared().currentUsername!, body: body, ext: [
                "agora_noti": "agora_addFriend"
            ])
            message.chatType = .groupChat
            message.isRead = true
            let conversation = AgoraChatClient.shared().chatManager?.getConversation(model.userId, type: .chat, createIfNotExist: true)
            conversation?.insert(message, error: nil)
            
            NotificationCenter.default.post(name: ConversationsUpdatedNotification, object: nil)
            NotificationCenter.default.post(name: GroupListChangedNotification, object: nil)
        } else {
            self.showAlert(message: "Agree to apply for failed".localized)
        }
        self.table.reloadData()
    }
    
    private func declineAction(model: AgoraApplyModel) {
        MBProgressHUD.showAdded(to: self.view, animated: true)
        switch model.type {
        case .contact(userId: let userId):
            AgoraChatClient.shared().contactManager?.declineFriendRequest(fromUser: userId, completion: { _, error in
                self.declineApplyFinished(model: model, error: error)
            })
        case .joinGroup(userId: let userId, groupId: let groupId):
            AgoraChatClient.shared().groupManager?.declineJoinGroupRequest(groupId, sender: userId, reason: nil, completion: { _, error in
                self.declineApplyFinished(model: model, error: error)
            })
        case .inviteGroup(userId: let userId, groupId: let groupId):
            AgoraChatClient.shared().groupManager?.declineGroupInvitation(groupId, inviter: userId, reason: nil, completion: { error in
                self.declineApplyFinished(model: model, error: error)
            })
        }
    }
    
    private func acceptAction(model: AgoraApplyModel) {
        MBProgressHUD.showAdded(to: self.view, animated: true)
        switch model.type {
        case .contact(userId: let userId):
            AgoraChatClient.shared().contactManager?.approveFriendRequest(fromUser: userId, completion: { _, error in
                self.acceptApplyFinished(model: model, error: error)
            })
        case .joinGroup(userId: let userId, groupId: let groupId):
            AgoraChatClient.shared().groupManager?.approveJoinGroupRequest(groupId, sender: userId, completion: { _, error in
                self.acceptApplyFinished(model: model, error: error)
            })
        case .inviteGroup(userId: let userId, groupId: let groupId):
            AgoraChatClient.shared().groupManager?.acceptInvitation(fromGroup: groupId, inviter: userId, completion: { _, error in
                self.acceptApplyFinished(model: model, error: error)
            })
        }
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.isSearchState ? self.searchResults.count : self.dataArray.count
    }

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "cell", for: indexPath)
        if let cell = cell as? ACDRequestCell {
            var applyModel: AgoraApplyModel?
            if self.isSearchState {
                applyModel = self.searchResults[indexPath.row] as? AgoraApplyModel
            } else {
                applyModel = self.dataArray[indexPath.row]
            }
            cell.model = applyModel
            cell.acceptHandle = { [unowned self] model in
                self.acceptAction(model: model)
            }
            cell.rejectHandle = { [unowned self] model in
                self.declineAction(model: model)
            }
        }
        return cell
    }
}
